File: bc2025/src/philip_06/utils/Random.cpp
#include "utils/Random.hpp"

namespace philip_06 {
namespace utils {

// philip_06.utils.Random is a tiny instance class (a 3-shift xorshift on a
// 64-bit seed). Both of its methods are one-liners that sit on the hot path
// (Constants::rng is queried every turn), so the bodies stay inline in
// Random.hpp per the performance rules in CONVENTIONS.md; this translation
// unit exists to keep the one-Java-file -> one-.hpp/.cpp-pair mapping and to
// compile-check the header on its own.
//
// Java semantics preserved in the header:
//   seed ^= (seed << 21);   ->  unsigned shift-left on the 64-bit seed
//   seed ^= (seed >>> 35);  ->  Java logical right shift, so the cast to
//                               uint64_t is required (C++ >> on a negative
//                               int64_t is arithmetic)
//   seed ^= (seed << 4);
//   return (int)(seed & 0x7FFFFFFF);  -> always non-negative, which is why
//                               nextBoolean()'s `% 2 == 0` needs no
//                               negative-remainder handling.
static_assert(sizeof(Random) == sizeof(int64_t), "Random is just its seed");

}  // namespace utils
}  // namespace philip_06

File: py/bc2025/src/philip_06/utils/Random.py
# Port of philip_06/utils/Random.java. See py/CONVENTIONS.md.
# philip_06.utils.Random is a tiny instance class (a 3-shift xorshift on a
# 64-bit seed) - NOT java.util.Random. Constants.rng is an instance of this.
from api import *

_MASK64 = 0xFFFFFFFFFFFFFFFF


class Random:
    def __init__(self, seed):
        self.seed = seed & _MASK64  # Java long

    def nextInt(self):
        # Java: seed ^= (seed << 21);  seed ^= (seed >>> 35);  seed ^= (seed << 4)
        # Java long arithmetic wraps at 64 bits; >>> is a logical right shift
        # (Python >> on a non-negative value is the same once masked).
        self.seed = (self.seed ^ ((self.seed << 21) & _MASK64)) & _MASK64
        self.seed = (self.seed ^ (self.seed >> 35)) & _MASK64
        self.seed = (self.seed ^ ((self.seed << 4) & _MASK64)) & _MASK64
        return self.seed & 0x7FFFFFFF

    def nextBoolean(self):
        # nextInt() is always non-negative, so % 2 matches Java exactly.
        return self.nextInt() % 2 == 0
